### DISCUSSION TITLE
How does switchMap cancel HTTP requests

### DISCUSSION THREAD

**Discussion by jtkb** [ReactiveX/rxjs#6975] [Q&A] [answered]
I'm curious as to how `switchMap` is able to cancel HTTP requests.

I know network requests using the [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) or [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest) APIs using [AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) but this necessitates setting up the request with the signal from an AbortController. So far when I have need to do network requests withing a switchMap I have used Angular's Http class and network requests are automatically cancelled if I pass new values to a switch map. But how is this achieved as I am not also passing in an Abortcontroller?
If for example I were to make a network request with the Fetch API directly but not use the Abortcontroller with is, would those network requests still be cancel-able?

**Comment by voliva** [comment, marked as answer]
`switchMap` just calls `.unsubscribe()` to the observable returned by the map function when a new value comes in.

If the observable returned by the map function cancels the HTTP request when calling `.unsubscribe()` on it, then the call will be canceled. If it doesn't, then it won't.

So I expect that `switchMap(() => fetch('https://...'))` will not cancel it - fetch returns a promise, which is not cancelable. It will get wrapped in an observable that on unsubscribe does nothing.

Angular's Http does internally use XHR to perform the actual request, and you can see it aborting the request in the teardown function of the observable here: https://github.com/angular/angular/blob/main/packages/common/http/src/xhr.ts#L317

If you wanted to use fetch in the same way, you would need to use an implementation that will cancel it on .unsubscribe(). It seems `rxjs` already has one implementation: https://rxjs.dev/api/fetch/fromFetch

**Reply by jtkb** [reply]
Thanks for your rely @voliva. What you stated confirms what I thought would be happening.

**Reply by baoqger** [reply]
Based on this post: https://stackoverflow.com/questions/24258279/internals-client-and-server-of-aborting-an-xmlhttprequest

abort() method will close the TCP connection.  I can't figure out how to reuse the TCP connection in this way.

**Reply by jakovljevic-mladen** [reply]
Unless XHR has option to keep connection open, which I doubt it has, there is nothing you can do in that case. However, I really think that modern browsers are smart enough so that they know how to keep connection open in an efficient way, so that developers don't have to worry about that.
